File: server/models/Problem.js
import mongoose from 'mongoose';

const inputSchema = new mongoose.Schema({
  value: { type: String, required: true },
  type: { 
    type: String, 
    enum: ['int', 'int[]', 'string', 'string[]', 'long', 'double'],
    required: true 
  },
  description: { type: String, default: '' }
});

const testCaseSchema = new mongoose.Schema({
  inputs: [inputSchema],
  expectedOutput: { type: String, default: '' },
  visibility: { type: String, enum: ['visible', 'hidden'], default: 'visible' }
});

const parameterSchema = new mongoose.Schema({
  name: { type: String, required: true },
  type: { type: String, required: true }
});

const functionSignatureSchema = new mongoose.Schema({
  functionName: { type: String, default: '' },
  returnType: { type: String, default: '' },
  parameters: [parameterSchema]
});

const codeTemplateSchema = new mongoose.Schema({
  python: { type: String, default: '' },
  java: { type: String, default: '' },
  cpp: { type: String, default: '' },
  c: { type: String, default: '' }
});

const exampleSchema = new mongoose.Schema({
  input: { type: String },
  output: { type: String },
  explanation: { type: String }
});

const statsSchema = new mongoose.Schema({
  likes: { type: Number, default: 0 },
  dislikes: { type: Number, default: 0 },
  acceptance: { type: String, default: '0%' },
  submissions: { type: String, default: '0' }
});

const problemSchema = new mongoose.Schema({
  serialNumber: { type: Number, required: true, unique: true },
  title: { type: String, required: true },
  difficulty: { 
    type: String, 
    enum: ['Easy', 'Medium', 'Hard'],
    required: true 
  },
  tags: [{ type: String }],
  companies: [{ type: String }],
  description: { type: String, required: true },
  examples: [exampleSchema],
  constraints: [{ type: String }],
  testCases: [testCaseSchema],
  codeTemplates: { type: codeTemplateSchema, default: {} },
  functionSignatures: {
    python: { type: functionSignatureSchema },
    java: { type: functionSignatureSchema },
    cpp: { type: functionSignatureSchema },
    c: { type: functionSignatureSchema }
  },
  stats: { type: statsSchema, default: {} },
  published: { type: Boolean, default: false },
  createdAt: { type: Date, default: Date.now },
  updatedAt: { type: Date, default: Date.now }
});

// Update the updatedAt timestamp before saving
problemSchema.pre('save', function(next) {
  this.updatedAt = Date.now();
  next();
});

const Problem = mongoose.model('Problem', problemSchema);

export default Problem;

File: backend/app/models/problem.py
from pydantic import BaseModel, Field
from typing import List, Optional, Dict
from datetime import datetime

class TestCase(BaseModel):
    inputs: List[str]
    expectedOutput: str
    visibility: str  # "visible" or "hidden"

class Parameter(BaseModel):
    name: str
    type: str

class FunctionSignature(BaseModel):
    functionName: str
    returnType: str
    parameters: List[Parameter]

class CodeTemplate(BaseModel):
    python: str = ""
    java: str = ""
    cpp: str = ""
    c: str = ""

class Example(BaseModel):
    input: str
    output: str
    explanation: Optional[str] = None

class ProblemStats(BaseModel):
    likes: int = 0
    dislikes: int = 0
    acceptance: str = "0%"
    submissions: str = "0"

class Problem(BaseModel):
    id: Optional[str] = Field(None, alias="_id")
    title: str
    difficulty: str  # "Easy", "Medium", "Hard"
    tags: List[str] = []
    companies: List[str] = []
    description: str
    examples: List[Example] = []
    constraints: List[str] = []
    testCases: List[TestCase] = []
    codeTemplates: CodeTemplate
    functionSignatures: Optional[Dict[str, FunctionSignature]] = None
    stats: ProblemStats = ProblemStats()
    published: bool = False
    createdAt: Optional[datetime] = None
    updatedAt: Optional[datetime] = None

    class Config:
        populate_by_name = True
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }

class ProblemCreate(BaseModel):
    title: str
    difficulty: str
    tags: List[str] = []
    companies: List[str] = []
    description: str
    examples: List[Example] = []
    constraints: List[str] = []
    testCases: List[TestCase] = []
    codeTemplates: CodeTemplate
    functionSignatures: Optional[Dict[str, FunctionSignature]] = None
    published: bool = False

class ProblemUpdate(BaseModel):
    title: Optional[str] = None
    difficulty: Optional[str] = None
    tags: Optional[List[str]] = None
    companies: Optional[List[str]] = None
    description: Optional[str] = None
    examples: Optional[List[Example]] = None
    constraints: Optional[List[str]] = None
    testCases: Optional[List[TestCase]] = None
    codeTemplates: Optional[CodeTemplate] = None
    functionSignatures: Optional[Dict[str, FunctionSignature]] = None
    stats: Optional[ProblemStats] = None
    published: Optional[bool] = None
